src/python/pants/backend/shell/lint/shellcheck/rules.py

```diff
@@ -0,0 +1,93 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from dataclasses import dataclass
+
+from pants.backend.shell.lint.shellcheck.subsystem import Shellcheck
+from pants.backend.shell.target_types import ShellSources
+from pants.core.goals.lint import LintRequest, LintResult, LintResults
+from pants.core.util_rules.external_tool import DownloadedExternalTool, ExternalToolRequest
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.fs import Digest, MergeDigests
+from pants.engine.platform import Platform
+from pants.engine.process import FallibleProcessResult, Process
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import Dependencies, DependenciesRequest, FieldSet, Sources, Targets
+from pants.engine.unions import UnionRule
+from pants.util.logging import LogLevel
+from pants.util.strutil import pluralize
+
+
+@dataclass(frozen=True)
+class ShellcheckFieldSet(FieldSet):
+    required_fields = (ShellSources,)
+
+    sources: ShellSources
+    dependencies: Dependencies
+
+
+class ShellcheckRequest(LintRequest):
+    field_set_type = ShellcheckFieldSet
+
+
+@rule(desc="Lint with Shellcheck", level=LogLevel.DEBUG)
```

Comment:
Its pre-existing, but the rule drop highlights this: It seems very strage that the "normal" way of doing things requires marking an @rule DEBUG and its process DEBUG when you'd think DEBUG is reserved for forays into -ldebug. I guess there is a conflation going on between what to display as persistent log lines vs what to ephemerally display in swim lanes and that's what makes the phrasing so odd.

Reply:
Agreed, and imo https://github.com/pantsbuild/pants/issues/7907 is still a problem. Rules can now have dynamic output upon completion, but their starting message is very vague because it can't introspect the input parameters.

I plan to revisit the conflation between swim lanes and log statements to fix the fallout from https://github.com/pantsbuild/pants/pull/11816. I'm not sure yet how much of a comprehensive fix I should make vs. something quick and dirty.